examples/step-78/step-78.cc

```diff
@@ -0,0 +1,835 @@
+// The program starts with the usual include files, all of which you should have
+// seen before by now:
+#include <deal.II/base/utilities.h>
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/base/function.h>
+#include <deal.II/base/logstream.h>
+#include <deal.II/lac/vector.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/dynamic_sparsity_pattern.h>
+#include <deal.II/lac/sparse_matrix.h>
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/precondition.h>
+#include <deal.II/lac/affine_constraints.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_refinement.h>
+#include <deal.II/grid/grid_out.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/grid/tria_iterator.h>
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_accessor.h>
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/vector_tools.h>
+#include <deal.II/numerics/error_estimator.h>
+#include <deal.II/numerics/solution_transfer.h>
+#include <deal.II/numerics/matrix_tools.h>
+#include <deal.II/numerics/data_out_stack.h>
+#include <deal.II/base/convergence_table.h>
+
+#include <fstream>
+#include <iostream>
+
+namespace BlackScholesSolver
+{
+  using namespace dealii;
+
+  template <int dim>
+  double coefficient(const Point<dim> &p)
+  {
+    return p.square();
+  }
+
+  template <int dim>
+  Tensor<1, dim> SCoefficient(const Point<dim> &p)
+  {
+    return Tensor<1, dim>(p);
+  }
+
+  /*
+  This Section creates a class for the known solution when testing using the MMS
+  I am using $v(\tau,S) = -\tau^2 -S^2 + 6$ for my solution
+  */
+  template <int dim>
+  class Solution : public Function<dim>
+  {
+  public:
+    Solution(double maturity_time);
+    virtual double value(const Point<dim> & p,
+                         const unsigned int component = 0) const override;
+
+    virtual Tensor<1, dim>
+    gradient(const Point<dim> & p,
+             const unsigned int component = 0) const override;
+
+    Tensor<1, dim> time_deriv(const Point<dim> & p,
+                              const unsigned int component = 0) const;
+
+  private:
+    double maturity_time;
+  };
+
+
+  template <int dim>
+  double Solution<dim>::value(const Point<dim> & p,
+                              const unsigned int component) const
+  {
+    return -std::pow(p(component), 2) - std::pow(this->get_time(), 2) + 6;
+  }
+
+  template <int dim>
+  Tensor<1, dim> Solution<dim>::gradient(const Point<dim> & p,
+                                         const unsigned int component) const
+  {
+    return Point<1>(-2 * p(component));
```

Comment:
Yes, this program only works for dim=1 for now. I added the assertion into the constructor for the BlackScholes class so that the program will fail immediately if the user tries to run any other dimension.

Reply:
Could you add this assertion also to the constructor of the `Solution` class? This class forms a closed entity that should also contain all of the assumptions you make.